R/ui_elements.R

```diff
@@ -11,8 +11,6 @@ use_telemetry <- function(id = "") {
   if (id != "" && !is.null(id)) {
     shiny_namespace <- glue::glue("{id}-")
   }
-
-
   shiny::tagList(
```

Comment:
Please use `shiny::singleton()` to return `shiny::tagList()` to prevent multiple erroneous `use_telemetry()`calls to additional tags in head.

Using `tags$head` would be nice to have, but not essential.

Reply:
See #159 on how I did that there

Reply:
Example below shows that when we call it 3 times it repeats the javascript code 3 times. With singleton it's only added once to DOM.

``` r
pkgload::load_all("/home/averissimo/work/opensource🟠/shiny.telemetry")
#> ℹ Loading shiny.telemetry

use_telemetry_as_is <- \() shiny.telemetry::use_telemetry()
use_telemetry_with_singleton <- \() shiny::singleton(use_telemetry())


shiny::tagList(
  use_telemetry_as_is(),
  use_telemetry_as_is(),
  use_telemetry_as_is()
) |> print()
#> <script type="text/javascript">
#>   $(document).on('shiny:sessioninitialized', function(event) {
#>     var br_ver = (function(){
#>       var ua= navigator.userAgent, tem, M;
#>       M = ua.match(/(opera|chrome|safari|firefox|msie|trident(?=\/))\/?\s*(\d+)/i) || [];
#>       if (/trident/i.test(M[1])) {
#>         tem = /\brv[ :]+(\d+)/g.exec(ua) || [];
#>         return 'IE '+(tem[1] || '');
#>       }
#>       if (M[1]=== 'Chrome') {
#>         tem= ua.match(/\b(OPR|Edge)\/(\d+)/);
#>         if(tem!= null) return tem.slice(1).join(' ').replace('OPR', 'Opera');
#>       }
#>       M = M[2]? [M[1], M[2]]: [navigator.appName, navigator.appVersion, '-?'];
#>       if((tem= ua.match(/version\/(\d+)/i))!= null) M.splice(1, 1, tem[1]);
#>       return M.join(' ');
#>     })();
#>       Shiny.setInputValue("browser_version", br_ver);
#>   });
#>       </script>
#> <script type="text/javascript">
#>   $(document).on('shiny:sessioninitialized', function(event) {
#>     var br_ver = (function(){
#>       var ua= navigator.userAgent, tem, M;
#>       M = ua.match(/(opera|chrome|safari|firefox|msie|trident(?=\/))\/?\s*(\d+)/i) || [];
#>       if (/trident/i.test(M[1])) {
#>         tem = /\brv[ :]+(\d+)/g.exec(ua) || [];
#>         return 'IE '+(tem[1] || '');
#>       }
#>       if (M[1]=== 'Chrome') {
#>         tem= ua.match(/\b(OPR|Edge)\/(\d+)/);
#>         if(tem!= null) return tem.slice(1).join(' ').replace('OPR', 'Opera');
#>       }
#>       M = M[2]? [M[1], M[2]]: [navigator.appName, navigator.appVersion, '-?'];
#>       if((tem= ua.match(/version\/(\d+)/i))!= null) M.splice(1, 1, tem[1]);
#>       return M.join(' ');
#>     })();
#>       Shiny.setInputValue("browser_version", br_ver);
#>   });
#>       </script>
#> <script type="text/javascript">
#>   $(document).on('shiny:sessioninitialized', function(event) {
#>     var br_ver = (function(){
#>       var ua= navigator.userAgent, tem, M;
#>       M = ua.match(/(opera|chrome|safari|firefox|msie|trident(?=\/))\/?\s*(\d+)/i) || [];
#>       if (/trident/i.test(M[1])) {
#>         tem = /\brv[ :]+(\d+)/g.exec(ua) || [];
#>         return 'IE '+(tem[1] || '');
#>       }
#>       if (M[1]=== 'Chrome') {
#>         tem= ua.match(/\b(OPR|Edge)\/(\d+)/);
#>         if(tem!= null) return tem.slice(1).join(' ').replace('OPR', 'Opera');
#>       }
#>       M = M[2]? [M[1], M[2]]: [navigator.appName, navigator.appVersion, '-?'];
#>       if((tem= ua.match(/version\/(\d+)/i))!= null) M.splice(1, 1, tem[1]);
#>       return M.join(' ');
#>     })();
#>       Shiny.setInputValue("browser_version", br_ver);
#>   });
#>       </script>

shiny::tagList(
  use_telemetry_with_singleton(),
  use_telemetry_with_singleton(),
  use_telemetry_with_singleton()
) |> print()
#> <script type="text/javascript">
#>   $(document).on('shiny:sessioninitialized', function(event) {
#>     var br_ver = (function(){
#>       var ua= navigator.userAgent, tem, M;
#>       M = ua.match(/(opera|chrome|safari|firefox|msie|trident(?=\/))\/?\s*(\d+)/i) || [];
#>       if (/trident/i.test(M[1])) {
#>         tem = /\brv[ :]+(\d+)/g.exec(ua) || [];
#>         return 'IE '+(tem[1] || '');
#>       }
#>       if (M[1]=== 'Chrome') {
#>         tem= ua.match(/\b(OPR|Edge)\/(\d+)/);
#>         if(tem!= null) return tem.slice(1).join(' ').replace('OPR', 'Opera');
#>       }
#>       M = M[2]? [M[1], M[2]]: [navigator.appName, navigator.appVersion, '-?'];
#>       if((tem= ua.match(/version\/(\d+)/i))!= null) M.splice(1, 1, tem[1]);
#>       return M.join(' ');
#>     })();
#>       Shiny.setInputValue("browser_version", br_ver);
#>   });
#>       </script>
```

<sup>Created on 2024-03-20 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>